interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,162 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - gadget
+    allow-connection:
+      plug-attributes:
+        shared-memory: $SLOT(shared-memory)
+    deny-auto-connection: true
+`
+
+// sharedMemoryInterface allows sharing sharedMemory between snaps
+type sharedMemoryInterface struct{}
+
+func (iface *sharedMemoryInterface) Name() string {
+	return "shared-memory"
+}
+
+func (iface *sharedMemoryInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              sharedMemorySummary,
+		BaseDeclarationSlots: sharedMemoryBaseDeclarationSlots,
+		AffectsPlugOnRefresh: true,
+	}
+}
+
+func (iface *sharedMemoryInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]
+	sharedMemory, ok := sharedMemoryAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,
+			slot.Attrs["shared-memory"])
+	}
+	if sharedMemory == "" {
+		if slot.Attrs == nil {
+			slot.Attrs = make(map[string]interface{})
+		}
+		// shared-memory defaults to "slot" name if unspecified
+		slot.Attrs["shared-memory"] = slot.Name
+	}
+
+	pathError := func(value interface{}) error {
+		return fmt.Errorf(`shared-memory "write" attribute must be a list of strings, not %v`, value)
+	}
+	// check that we have a write path
+	writeAttr, ok := slot.Attrs["write"]
+	if !ok {
+		return pathError("unset")
+	}
+
+	pathList, ok := writeAttr.([]interface{})
+	if !ok || len(pathList) == 0 {
+		return pathError(pathList)
+	}
+
+	for _, pathAttr := range pathList {
+		path, ok := pathAttr.(string)
+		if !ok {
+			return pathError(pathList)
+		}
+		path = strings.TrimSpace(path)
+		if len(path) == 0 {
+			return fmt.Errorf("shared-memory interface path is empty")
+		}
+		if err := apparmor.ValidateNoAppArmorRegexp(path); err != nil {
+			return fmt.Errorf("shared-memory interface path is invalid: %v", err)
+		}
+
+		if !cleanSubPath(path) {
+			return fmt.Errorf("shared-memory interface path is not clean: %q", path)
+		}
+	}
+	return nil
+}
+
+func writeSharedMemoryPaths(w io.Writer, slot *interfaces.ConnectedSlot) {
+	var pathList []interface{}
+	_ = slot.Attr("write", &pathList)
+	for _, pathAttr := range pathList {
+		// All checks were already done in BeforePrepare{Plug,Slot}
+		path, _ := pathAttr.(string)
+		path = strings.TrimSpace(path)
+		// Ubuntu 14.04 uses /run/shm instead of the most common /dev/shm
+		fmt.Fprintf(w, "\"/{dev,run}/shm/%s\" mrwk,\n", path)
```

Comment:
also this does not permit the slot to specify both the short hand and the long hand, i.e. both

```yaml
slots:
  foo-mem:
    interface: shared-memory
    write: [ /dev/shm/org.foo.mem ]
```

and 

```yaml
slots:
  foo-mem:
    interface: shared-memory
    write: [ org.foo.mem ]
```

Reply:
Do we already know that we'll want to support paths other than those used by `shm_open()`? If not, I'd rather leave this to a follow-up PR, to be made when the use-case is clear.

Reply:
The point here is not about supporting paths outside of `/dev/shm/`, the point is about accepting both forms of the interface definition from the spec (both of my examples above should result in the same policy written out to disk). Ideally we would validate that the long-form specification only has `/dev/shm` as the prefix in the long-hand. You could switch on if the first character is `/` or not. 

I personally don't think this should be a huge change to support both the absolute and the relative paths, but I do think if you only want to support the short-hand in this PR and add support for the long-hand in another PR that there should be an error for the absolute path case which AFAICT there is not today

Reply:
I see. Well, since there is a small window for further discussions (like, should we also allow `/run/shm/...`?), I'll just take the second option and add an error code if the path contains a `/`.